## Import packages and read in data

In [68]:
# import required packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

In [69]:
# read in data for test dataframe
test_df = pd.read_csv('clean_test_dataframe.csv')
test_df.head()

Additional_Number_of_Scoring  Average_Score  \
0                          1005            8.3   
1                           487            8.8   
2                           570            9.3   
3                          2623            8.7   
4                           365            8.1   

   Review_Total_Negative_Word_Counts  Total_Number_of_Reviews  \
0                                  0                     3469   
1                                 46                     1805   
2                                  0                     2319   
3                                 19                    12158   
4                                 13                     1567   

   Review_Total_Positive_Word_Counts  \
0                                  6   
1                                121   
2                                  9   
3                                  0   
4                                  0   

   Total_Number_of_Reviews_Reviewer_Has_Given  days_since_review        lat  \
0                                           2                297  51.559096   
1                                           1                  4  51.474469   
2                                           1                590  51.505696   
3                                           2                269  51.500961   
4                                           2                373  51.515125   

        lng  weekday_of_review  ...  n_worry  n_worth  n_would  n_write  \
0 -0.284704                  0  ...        0        0        0        0   
1 -0.019428                  6  ...        0        0        0        0   
2 -0.101525                  1  ...        0        0        0        0   
3 -0.116591                  0  ...        0        0        0        0   
4 -0.160066                  1  ...        0        0        0        0   

   n_wrong  n_year  n_yes  n_yet  n_young  rating  
0        0       0      0      0        0       0  
1        0       0      0      0        0       1  
2        0       0      0      0        0       1  
3        0       0      0      0        0       0  
4        0       0      0      0        0       0  

[5 rows x 2744 columns]

In [82]:
test_df.describe()

Additional_Number_of_Scoring  Average_Score  \
count                   4267.000000    4267.000000   
mean                     746.518163       8.347082   
std                      609.426996       0.602030   
min                       18.000000       6.400000   
25%                      328.000000       8.000000   
50%                      541.000000       8.400000   
75%                      974.000000       8.800000   
max                     2682.000000       9.600000   

       Review_Total_Negative_Word_Counts  Total_Number_of_Reviews  \
count                        4267.000000              4267.000000   
mean                           19.422311              3032.284040   
std                            29.274398              2424.732703   
min                             0.000000                59.000000   
25%                             2.000000              1402.500000   
50%                            10.000000              2312.000000   
75%                            25.000000              3826.000000   
max                           359.000000             12158.000000   

       Review_Total_Positive_Word_Counts  \
count                        4267.000000   
mean                           16.260136   
std                            19.150631   
min                             0.000000   
25%                             5.000000   
50%                            10.000000   
75%                            21.000000   
max                           371.000000   

       Total_Number_of_Reviews_Reviewer_Has_Given  days_since_review  \
count                                 4267.000000        4267.000000   
mean                                     4.919147         355.043356   
std                                      7.685281         202.717054   
min                                      1.000000           0.000000   
25%                                      1.000000         185.000000   
50%                                      2.000000         358.000000   
75%                                      5.000000         527.000000   
max                                     92.000000         730.000000   

               lat          lng  weekday_of_review  ...      n_worry  \
count  4267.000000  4267.000000        4267.000000  ...  4267.000000   
mean     51.511344    -0.134247           2.589407  ...     0.000469   
std       0.018176     0.071819           2.095507  ...     0.021647   
min      51.420266    -0.369758           0.000000  ...     0.000000   
25%      51.500961    -0.181389           1.000000  ...     0.000000   
50%      51.510841    -0.136603           2.000000  ...     0.000000   
75%      51.517972    -0.099716           4.000000  ...     0.000000   
max      51.606805     0.059042           6.000000  ...     1.000000   

           n_worth      n_would      n_write      n_wrong       n_year  \
count  4267.000000  4267.000000  4267.000000  4267.000000  4267.000000   
mean      0.006093     0.072416     0.000703     0.004453     0.005625   
std       0.077830     0.304915     0.026509     0.070020     0.083670   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.000000     0.000000     0.000000   
max       1.000000     4.000000     1.000000     2.000000     2.000000   

             n_yes        n_yet      n_young       rating  
count  4267.000000  4267.000000  4267.000000  4267.000000  
mean      0.001172     0.002578     0.002578     0.576049  
std       0.034215     0.055143     0.055143     0.494241  
min       0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     1.000000  
75%       0.000000     0.000000     0.000000     1.000000  
max       1.000000     2.000000     2.000000     1.00000

In [70]:
# read in data for train dataframe
train_df = pd.read_csv('clean_train_dataframe.csv')
train_df.head()

Additional_Number_of_Scoring  Average_Score  \
0                           620            9.0   
1                          1258            9.4   
2                           995            8.1   
3                           853            8.4   
4                          1243            8.1   

   Review_Total_Negative_Word_Counts  Total_Number_of_Reviews  \
0                                  0                     1974   
1                                  6                     4204   
2                                  2                     3826   
3                                  7                     2726   
4                                 11                     6608   

   Review_Total_Positive_Word_Counts  \
0                                164   
1                                  4   
2                                 38   
3                                 10   
4                                  8   

   Total_Number_of_Reviews_Reviewer_Has_Given  days_since_review        lat  \
0                                           1                562  51.506558   
1                                           5                276  51.502435   
2                                           1                129  51.504348   
3                                          10                164  51.507377   
4                                          69                639  51.513556   

        lng  weekday_of_review  ...  n_worry  n_worth  n_would  n_write  \
0 -0.004514                  1  ...        0        0        0        0   
1 -0.000250                  0  ...        0        0        0        0   
2 -0.033444                  0  ...        0        0        0        0   
3  0.038657                  0  ...        0        0        0        0   
4 -0.180002                  1  ...        0        0        0        0   

   n_wrong  n_year  n_yes  n_yet  n_young  rating  
0        0       0      0      0        0       1  
1        0       0      0      0        0       1  
2        0       0      0      0        0       1  
3        0       0      0      0        0       0  
4        0       0      0      0        0       0  

[5 rows x 2744 columns]

## Modeling

### Q1: Employ a linear classifier on this dataset:

### A: Fit a logistic regression model to this data with the solver set to `lbfgs`. What is the accuracy score on the test set?

We will first set up X and y variables based on `rating` being the target column on the training set first.

In [71]:
X_train = train_df.drop(columns="rating")
y_train = train_df["rating"]

In [72]:
X_train.shape

(12798, 2743)

In [73]:
y_train.shape

(12798,)

Now that we have created the training set, we will fit a logistic regression  with solver set to `lbfgs`.

In [74]:
# instantiate model
train_logit = LogisticRegression(solver='lbfgs', random_state=1)

#fit model
train_logit.fit(X_train, y_train);

C:\Users\avitr\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [76]:
train_logit.intercept_

array([-9.84203399e-07])

In [77]:
train_logit.coef_

array([[ 4.54565764e-04,  7.59844720e-02, -4.69075885e-02, ...,
         5.22795186e-04, -4.34995162e-06,  2.79403969e-04]])

In [44]:
# score model
print(f'Score on train: {train_logit.score(X_train, y_train)}')


Score on train: 0.7238631036099391


We will now fit the logistic regression with solver set to `lbfgs` to the test set. Like the training set, the test set's variables will be based on `rating` being the target column.

In [45]:
X_test = test_df.drop(columns="rating")
y_test = test_df["rating"]

In [46]:
# instantiate model
test_logit = LogisticRegression(solver='lbfgs', random_state=1)

#fit model
test_logit.fit(X_test, y_test);

C:\Users\avitr\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [47]:
# score model
print(f'Score on test: {test_logit.score(X_test, y_test)}')


Score on test: 0.720881181157722


The accuracy score on the test set is 0.7190063276306539. This is lower than the accuracy score on the training set which is at 0.7238631036099391.

### B: What are the 20 words most predictive of a good review (from the positive review column)? What are the 20 words most predictive with a bad review (from the negative review column)? Use the regression coefficients to answer this question.

#### we need to create a new test and train dataframe that exclude hotel name columns
#### then we need to run new X/y_test and X/y_train sets
#### then we run bag of words/count vectorizer to get the count of coefficients....

First, we will make a document-term matrix that's based on the training data.

In the chart above, we can see that the top 20 list contains stopwords. Further, none of the words labeled with a 'p_' or 'n_', indicating association with positive or negative reviews, respectively, are not present.

Given this, we need to remove stop words from the `bagofwords`. After doing this, we will need to use a Logistic Regression model to find the top 20 terms most preditive for positive reviews, and the top 20 terms most predictive for negative reviews.

Now, we will use get coefficients generated by the Logistic Regression model used in Q1 to determine the top 20 words that are most predictive of a positive review, and the top 20 words that are most predictive of a negative review.